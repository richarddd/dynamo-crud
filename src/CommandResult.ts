export class CommandResult<Data, RawResponse> {
  public readonly data: Data
  public readonly rawResponse: RawResponse
  constructor(data: Data, rawResponse: RawResponse) {
    this.data = data
    this.rawResponse = rawResponse
  }
}
