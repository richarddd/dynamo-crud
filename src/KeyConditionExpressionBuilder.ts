import { marshall } from '@aws-sdk/util-dynamodb'
import { GetPrimaryKey, PrimaryKeyNameCandidates } from './types'

type ComparisonOperator = '=' | '<' | '<=' | '>' | '>=' | 'BETWEEN' | 'begins_with'

export class KeyConditionExpressionBuilder<
  Model extends Record<string, any>,
  PK extends PrimaryKeyNameCandidates<Model>,
  SK extends PrimaryKeyNameCandidates<Model> = never
> {
  public readonly partitionKeyName: PK
  public readonly partitionKeyValue: Model[PK]

  private sortKeyCondition: string | undefined
  private sortKeyName: SK | undefined
  private sortKeyValue1: Model[SK] | undefined
  private sortKeyValue2: Model[SK] | undefined

  /**
   * @param primaryKey [Note] PK should be the first key of the object.
   */
  constructor(primaryKey: GetPrimaryKey<Model, PK, SK, false>) {
    const [pk, sk] = Object.entries(primaryKey)

    this.partitionKeyName = pk[0] as PK
    this.partitionKeyValue = pk[1]

    if (sk) {
      this.where(sk[0] as SK, '=', sk[1])
    }
  }

  where(
    sortKeyName: SK,
    comparisonOperator: ComparisonOperator,
    ...sortKeyValues: ComparisonOperator extends 'BETWEEN' ? [Model[SK], Model[SK]] : [Model[SK]]
  ): this {
    this.sortKeyName = sortKeyName
    this.sortKeyValue1 = sortKeyValues[0]
    this.sortKeyValue2 = '1' in sortKeyValues ? sortKeyValues[1] : undefined

    if (comparisonOperator === 'BETWEEN') {
      this.sortKeyCondition = `#sk BETWEEN :sv AND :sv2`

      return this
    }

    if (comparisonOperator === 'begins_with') {
      this.sortKeyCondition = `begins_with( #sk, :sv )`

      return this
    }

    this.sortKeyCondition = `#sk ${comparisonOperator} :sv`

    return this
  }

  getSortKeyName() {
    return this.sortKeyName
  }

  compile() {
    const KeyConditionExpression =
      '#pk = :pv' + (this.sortKeyCondition ? ` AND ${this.sortKeyCondition}` : '')

    const ExpressionAttributeValues: { ':pv': any; ':sv'?: any; ':sv2'?: any } = {
      ':pv': this.partitionKeyValue,
    }
    if (this.sortKeyValue1) {
      ExpressionAttributeValues[':sv'] = this.sortKeyValue1
    }
    if (this.sortKeyValue2) {
      ExpressionAttributeValues[':sv2'] = this.sortKeyValue2
    }

    const ExpressionAttributeNames: {
      '#pk': PK
      '#sk'?: SK
    } = {
      '#pk': this.partitionKeyName,
    }
    if (this.sortKeyName) {
      ExpressionAttributeNames['#sk'] = this.sortKeyName
    }

    return {
      KeyConditionExpression,
      ExpressionAttributeNames,
      ExpressionAttributeValues: marshall(ExpressionAttributeValues),
    }
  }
}
