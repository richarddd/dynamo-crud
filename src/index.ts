export * from './GetItemCommand'
export * from './QueryCommand'
export * from './UpdateItemCommand'
export * from './DeleteItemCommand'
export * from './PutItemCommand'
export * from './CommandResult'
export * from './UpdateExpressionBuilder'
export * from './KeyConditionExpressionBuilder'

export * from './types'
export * from './DynamoTable'
