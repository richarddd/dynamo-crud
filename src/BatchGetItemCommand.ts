import { PrimaryKeyCandidates, TokenBucket } from './types'
import { CommandResult } from './CommandResult'
import { marshall, unmarshall } from '@aws-sdk/util-dynamodb'
import { DynamoDB, KeysAndAttributes, BatchGetItemCommandInput } from '@aws-sdk/client-dynamodb'
import { BatchGetItemCommandStream } from './stream'
import { selectAttributes } from './utils'

export class BatchGetItemCommand<Model extends Record<string, any>> {
  private readonly client: DynamoDB
  private readonly tableName: string
  private selectedAttributes?: (keyof Model)[]
  private readonly primaryKeys: PrimaryKeyCandidates<Model>[]
  private batchGetItemCommandInput: Partial<BatchGetItemCommandInput> = {}
  public readonly tokenBucket?: TokenBucket

  constructor(args: {
    client: DynamoDB
    tableName: string
    primaryKeys: PrimaryKeyCandidates<Model>[]
    tokenBucket?: TokenBucket
  }) {
    this.client = args.client
    this.tableName = args.tableName
    this.primaryKeys = args.primaryKeys
    this.tokenBucket = args.tokenBucket
  }

  setCommandInput(options: Partial<BatchGetItemCommandInput>): this {
    this.batchGetItemCommandInput = { ...this.batchGetItemCommandInput, ...options }
    return this
  }

  getCommandInput() {
    return this.batchGetItemCommandInput
  }

  select(...attributeNames: (keyof Model)[]): this {
    this.selectedAttributes = attributeNames

    return this
  }

  createReadableStream() {
    return new BatchGetItemCommandStream<Model>({
      command: this,
    })
  }

  async run() {
    return this.createReadableStream().consume()
  }

  async _run({ keysAndAttributes }: { keysAndAttributes: KeysAndAttributes }) {
    const rawResponse = await this.client.batchGetItem({
      RequestItems: {
        [this.tableName]: keysAndAttributes,
      },
      ...this.batchGetItemCommandInput,
      ...(this.tokenBucket ? { ReturnConsumedCapacity: 'TOTAL' } : {}),
    })

    const items =
      rawResponse?.Responses && rawResponse.Responses[this.tableName]
        ? (unmarshall(rawResponse.Responses[this.tableName] as any) as Model[])
        : undefined

    return new CommandResult(items, rawResponse)
  }

  getKeysAndAttributes(): KeysAndAttributes {
    const Keys = this.primaryKeys.map((primaryKey) => {
      return marshall(primaryKey)
    })

    const keysAndAttributes: KeysAndAttributes = {
      Keys,
    }

    selectAttributes(keysAndAttributes, this.selectedAttributes as string[])

    return keysAndAttributes
  }
}
