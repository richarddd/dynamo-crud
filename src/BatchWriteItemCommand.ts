import { PrimaryKeyCandidates, TokenBucket } from './types'
import { CommandResult } from './CommandResult'
import { marshall } from '@aws-sdk/util-dynamodb'
import { DynamoDB, BatchWriteItemCommandInput, WriteRequest } from '@aws-sdk/client-dynamodb'
import { BatchWriteItemCommandStream } from './stream'

export class BatchWriteItemCommand<Model extends Record<string, any>> {
  private readonly client: DynamoDB
  private readonly tableName: string
  private readonly itemsPut?: Model[]
  private readonly itemsDelete?: PrimaryKeyCandidates<Model>[]
  public readonly tokenBucket?: TokenBucket
  private batchWriteItemCommandInput: Partial<BatchWriteItemCommandInput> = {}

  constructor(args: {
    client: DynamoDB
    tableName: string
    itemsPut?: Model[]
    itemsDelete?: PrimaryKeyCandidates<Model>[]
    tokenBucket?: TokenBucket
  }) {
    this.client = args.client
    this.tableName = args.tableName
    this.itemsPut = args.itemsPut
    this.itemsDelete = args.itemsDelete
    this.tokenBucket = args.tokenBucket
  }

  setCommandInput(options: Partial<BatchWriteItemCommandInput>): this {
    this.batchWriteItemCommandInput = { ...this.batchWriteItemCommandInput, ...options }
    return this
  }

  createReadableStream() {
    return new BatchWriteItemCommandStream<Model>({
      command: this,
    })
  }

  async run() {
    return this.createReadableStream().consume()
  }

  async _run({ writeRequests }: { writeRequests: WriteRequest[] }) {
    const rawResponse = await this.client.batchWriteItem({
      RequestItems: {
        [this.tableName]: writeRequests,
      },
      ...this.batchWriteItemCommandInput,
      ...(this.tokenBucket ? { ReturnConsumedCapacity: 'TOTAL' } : {}),
    })

    return new CommandResult(undefined, rawResponse)
  }

  getWriteRequests(): WriteRequest[] {
    if (this.itemsPut) {
      return this.itemsPut.map((item) => {
        return {
          PutRequest: { Item: marshall(item) },
        }
      })
    }

    if (this.itemsDelete) {
      return this.itemsDelete.map((primaryKey) => {
        return {
          DeleteRequest: { Key: marshall(primaryKey) },
        }
      })
    }

    throw new Error(`[dynamo-crud] Couldn't find 'itemsPut' or 'itemsDelete'`)
  }
}
