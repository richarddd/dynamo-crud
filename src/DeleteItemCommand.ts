import { CommandResult } from './CommandResult'
import { marshall } from '@aws-sdk/util-dynamodb'
import { DeleteItemCommandInput, DynamoDB } from '@aws-sdk/client-dynamodb'
import { PrimaryKeyCandidates, TokenBucket } from './types'

export class DeleteItemCommand<Model extends Record<string, any>> {
  private readonly client: DynamoDB
  private readonly tableName: string
  private readonly primaryKey: PrimaryKeyCandidates<Model>
  private readonly tokenBucket?: TokenBucket
  private deleteItemCommandInput: Partial<DeleteItemCommandInput> = {}

  constructor(args: {
    client: DynamoDB
    tableName: string
    primaryKey: PrimaryKeyCandidates<Model>
    tokenBucket?: TokenBucket
  }) {
    this.client = args.client
    this.tableName = args.tableName
    this.primaryKey = args.primaryKey
    this.tokenBucket = args.tokenBucket
  }

  setCommandInput(deleteItemCommandInput: Partial<DeleteItemCommandInput>): this {
    this.deleteItemCommandInput = { ...this.deleteItemCommandInput, ...deleteItemCommandInput }
    return this
  }

  async run() {
    const rawResponse = await this.client.deleteItem({
      TableName: this.tableName,
      Key: marshall(this.primaryKey),
      ...this.deleteItemCommandInput,
      ...(this.tokenBucket ? { ReturnConsumedCapacity: 'TOTAL' } : {}),
    })

    if (this.tokenBucket) {
      // TODO move to head
      const consumedCapacityUnits = rawResponse.ConsumedCapacity!.CapacityUnits!
      this.tokenBucket.removeTokens(consumedCapacityUnits)
    }

    return new CommandResult(undefined, rawResponse)
  }
}
