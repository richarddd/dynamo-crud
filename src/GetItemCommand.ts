import { CommandResult } from './CommandResult'
import { marshall, unmarshall } from '@aws-sdk/util-dynamodb'
import { DynamoDB, GetItemCommandInput } from '@aws-sdk/client-dynamodb'
import { PrimaryKeyCandidates } from './types'

export class GetItemCommand<Model extends Record<string, any>> {
  private readonly client: DynamoDB
  private readonly tableName: string
  private readonly primaryKey: PrimaryKeyCandidates<Model>
  private getItemCommandInput: Partial<GetItemCommandInput> = {}

  constructor(args: { client: DynamoDB; tableName: string; primaryKey: PrimaryKeyCandidates<Model> }) {
    this.client = args.client
    this.tableName = args.tableName
    this.primaryKey = args.primaryKey
  }

  setCommandInput(getItemCommandInput: Partial<GetItemCommandInput>): this {
    this.getItemCommandInput = { ...this.getItemCommandInput, ...getItemCommandInput }
    return this
  }

  select(...attributeNames: (keyof Model)[]): this {
    return this.setCommandInput({
      ProjectionExpression: attributeNames.join(', '),
    })
  }

  async run() {
    const rawResponse = await this.client.getItem({
      TableName: this.tableName,
      Key: marshall(this.primaryKey),
      ...this.getItemCommandInput,
    })
    const data = rawResponse.Item ? (unmarshall(rawResponse.Item) as Model) : undefined

    return new CommandResult(data, rawResponse)
  }
}
