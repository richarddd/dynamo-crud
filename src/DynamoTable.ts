import { GetItemCommand } from './GetItemCommand'
import { PutItemCommand } from './PutItemCommand'
import { DynamoDB } from '@aws-sdk/client-dynamodb'
import { UpdatableScanCommand } from './ScanCommand'
import { UpdatableQueryCommand } from './QueryCommand'
import { DeleteItemCommand } from './DeleteItemCommand'
import { UpdateItemCommand } from './UpdateItemCommand'
import { BatchGetItemCommand } from './BatchGetItemCommand'
import { BatchWriteItemCommand } from './BatchWriteItemCommand'
import { GetPrimaryKey, PrimaryKeyNameCandidates, TokenBucket } from './types'

export class DynamoTable<
  Model extends Record<string, any>,
  BasePK extends PrimaryKeyNameCandidates<Model>,
  BaseSK extends PrimaryKeyNameCandidates<Model> = never
> {
  private readonly client: DynamoDB
  public readonly tableName: string
  public readonly partitionKey: BasePK
  public readonly sortKey?: BaseSK
  private readonly tokenBucket?: TokenBucket

  constructor(args: {
    client: DynamoDB
    tableName: string
    partitionKey: BasePK
    sortKey?: BaseSK
    tokenBucket?: TokenBucket
  }) {
    this.client = args.client
    this.tableName = args.tableName
    this.partitionKey = args.partitionKey
    this.sortKey = args.sortKey
    this.tokenBucket = args.tokenBucket
  }

  scan() {
    const { client, tableName, tokenBucket, partitionKey: basePartitionKey, sortKey: baseSortKey } = this

    return new UpdatableScanCommand<Model>({
      client,
      tableName,
      basePartitionKey,
      baseSortKey,
      tokenBucket,
    })
  }

  update(primaryKey: GetPrimaryKey<Model, BasePK, BaseSK>) {
    const { client, tableName, tokenBucket } = this

    return new UpdateItemCommand<Model>({
      client,
      tableName,
      primaryKey: primaryKey as any,
      tokenBucket,
    })
  }

  query<IndexPrimaryKeyNames extends PrimaryKeyNameCandidates<Model>>(
    primaryKey: { [K in IndexPrimaryKeyNames]: Model[K] }
  ) {
    const { client, tableName, partitionKey: basePartitionKey, sortKey: baseSortKey, tokenBucket } = this
    return new UpdatableQueryCommand<Model, IndexPrimaryKeyNames, IndexPrimaryKeyNames>({
      client,
      tableName,
      primaryKey: primaryKey as any,
      basePartitionKey,
      baseSortKey,
      tokenBucket,
    })
  }

  find<PrimaryKey extends GetPrimaryKey<Model, BasePK, BaseSK> | GetPrimaryKey<Model, BasePK, BaseSK>[]>(
    primaryKey: PrimaryKey
  ): PrimaryKey extends GetPrimaryKey<Model, BasePK, BaseSK>[]
    ? BatchGetItemCommand<Model>
    : GetItemCommand<Model> {
    const { client, tableName, tokenBucket } = this

    if (Array.isArray(primaryKey)) {
      return new BatchGetItemCommand({ client, tableName, primaryKeys: primaryKey, tokenBucket }) as any
    }

    return new GetItemCommand({
      client,
      tableName,
      primaryKey: primaryKey as any,
      tokenBucket,
    }) as any
  }

  delete<PrimaryKey extends GetPrimaryKey<Model, BasePK, BaseSK> | GetPrimaryKey<Model, BasePK, BaseSK>[]>(
    primaryKey: PrimaryKey
  ): PrimaryKey extends GetPrimaryKey<Model, BasePK, BaseSK>[]
    ? BatchWriteItemCommand<Model>
    : DeleteItemCommand<Model> {
    const { client, tableName, tokenBucket } = this

    if (Array.isArray(primaryKey)) {
      return new BatchWriteItemCommand({ client, tableName, itemsDelete: primaryKey, tokenBucket }) as any
    }

    return new DeleteItemCommand({
      client,
      tableName,
      primaryKey: primaryKey as any,
      tokenBucket,
    }) as any
  }

  put<Item extends Model | Model[]>(
    item: Item
  ): Item extends Model[] ? BatchWriteItemCommand<Model> : PutItemCommand<Model> {
    const { client, tableName, tokenBucket } = this

    if (Array.isArray(item)) {
      return new BatchWriteItemCommand({ client, tableName, itemsPut: item, tokenBucket }) as any
    }

    return new PutItemCommand({
      client,
      tableName,
      item,
      tokenBucket,
    }) as any
  }
}
