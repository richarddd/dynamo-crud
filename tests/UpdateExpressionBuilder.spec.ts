import { PostModel } from './data'
import { UpdateExpressionBuilder } from '../src'

describe('KeyConditionExpressionBuilder', () => {
  const builder = new UpdateExpressionBuilder<PostModel>()

  it('build update expression', () => {
    expect(
      builder
        .set('title', 'PostTitle')
        .set('tags', ['tag1', 'tag2'])
        .set('author.name', 'foobar')
        .setCalc('likeCount', 'id', '+', 'author.id')
        .setIfNotExists('author.name', 'foobar')
        .setListAppend('tags', ['tag-1', 'tag-2'])
        .delete('categories', new Set(['cat-1']))
        .remove('author.followers.0.name')
        .add('likeCount', 1)
        .add('comments.2.likeCount', 5)
        .compile()
    ).toMatchSnapshot()
  })
})
