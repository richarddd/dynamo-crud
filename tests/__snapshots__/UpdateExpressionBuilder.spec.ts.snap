// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KeyConditionExpressionBuilder build update expression 1`] = `
Object {
  "ExpressionAttributeNames": Object {
    "#p0": "title",
    "#p1": "tags",
    "#p10": "tags",
    "#p11": "author",
    "#p12": "followers",
    "#p13": "0",
    "#p14": "name",
    "#p15": "likeCount",
    "#p16": "comments",
    "#p17": "2",
    "#p18": "likeCount",
    "#p19": "categories",
    "#p2": "author",
    "#p3": "name",
    "#p4": "likeCount",
    "#p5": "id",
    "#p6": "author",
    "#p7": "id",
    "#p8": "author",
    "#p9": "name",
  },
  "ExpressionAttributeValues": Object {
    ":v0": Object {
      "S": "PostTitle",
    },
    ":v1": Object {
      "L": Array [
        Object {
          "S": "tag1",
        },
        Object {
          "S": "tag2",
        },
      ],
    },
    ":v2": Object {
      "S": "foobar",
    },
    ":v3": Object {
      "S": "foobar",
    },
    ":v4": Object {
      "L": Array [
        Object {
          "S": "tag-1",
        },
        Object {
          "S": "tag-2",
        },
      ],
    },
    ":v5": Object {
      "N": "1",
    },
    ":v6": Object {
      "N": "5",
    },
    ":v7": Object {
      "SS": Array [
        "cat-1",
      ],
    },
  },
  "UpdateExpression": "SET #p0 = :v0, #p1 = :v1, #p2.#p3 = :v2, #p4 = #p5 + #p6.#p7, #p8.#p9 = if_not_exists(#p8.#p9, :v3), #p10 = list_append(#p10, :v4)
REMOVE #p11.#p12.#p13.#p14
ADD #p15 :v5, #p16.#p17.#p18 :v6
DELETE #p19 :v7",
}
`;
